src/threadutil.cpp

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <atomic>
+#include <thread>
+
+#include <threadutil.h>
+
+/*
+ * TODO: using thread_local changes the abi in ways that may not play nice
+ * when the c++ stdlib is linked dynamically. Disable it until thorough
+ * testing has been done.
+ *
+ * mingw32's implementation of thread_local has also been shown to behave
+ * erroneously under concurrent usage; see:
+ *
+ *   https://gist.github.com/jamesob/fe9a872051a88b2025b1aa37bfa98605
+ */
+#undef HAVE_THREAD_LOCAL
+
+#ifdef HAVE_SYS_PRCTL_H
+#include <sys/prctl.h> // For prctl, PR_SET_NAME, PR_GET_NAME
+#endif
+
+#if (defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__DragonFly__))
+#include <pthread_np.h>
+#endif
+
+#if defined(HAVE_THREAD_LOCAL)
+
+#elif defined(HAVE_PTHREAD)
+#include <pthread.h>
+
+#else
+#include <mutex>
+#include <unordered_map>
+#endif
+
+struct thread_data_type
+{
+    constexpr thread_data_type() = default;
+    thread_data_type(long id, std::string name) : m_id(id), m_name(std::move(name)){}
+
+    long m_id{0};
+    std::string m_name{""};
+};
+
+void thread_util::SetProcessName(const char* name)
+{
+#if defined(PR_SET_NAME)
+    // Only the first 15 characters are used (16 - NUL terminator)
+    ::prctl(PR_SET_NAME, name, 0, 0, 0);
+#elif (defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__DragonFly__))
+    pthread_set_name_np(pthread_self(), name);
+#elif defined(MAC_OSX)
+    pthread_setname_np(name);
+#else
+    // Prevent warnings for unused parameters...
+    (void)name;
+#endif
+}
+
+std::string thread_util::GetProcessName()
+{
+#if !defined(CAN_READ_PROCESS_NAME)
+    return "";
+#endif
+
+    char threadname_buff[16];
+    char* pthreadname_buff = (char*)(&threadname_buff);
```

Comment:
I don't think you need this variable; using `threadname_buff` will automatically decay to `(char *)`.